Optional commit spell checking

Hovemeyer5 opened the pull request:
Spell checking in commit summary & description #1597
Closes #1597

## Description
Made spell checking enabled on commit summary and message by default.

From commit summary or description context menu, user can disable or enable spell checking. 

If enabled, summary and description of commit use spellchecker. 

### Screenshots

![image](https://user-images.githubusercontent.com/7199430/98481215-ceb2d600-21c6-11eb-92b8-474f7b801bda.png)


## Release notes
Notes: "Enable spellcheck on commit summary and description."

Hovemeyer5 committed:
Enable spellcheck

Hovemeyer5 committed:
Add spellcheck option to commit inputs

Hovemeyer5 committed:
Add spell check option

Hovemeyer5 committed:
Add to app store to set in  local store

Hovemeyer5 committed:
Hook up in app state

Hovemeyer5 committed:
Pass down to commit message

Hovemeyer5 committed:
make other text box not use spellcheck

Hovemeyer5 committed:
Remove unrelated message

Hovemeyer5 committed:
Prettier

niik commented:
Hi @Hovemeyer5, thanks so much for opening this and for wanting to contribute!

The changes look very idiomatic and well laid out and I especially appreciate the attention given to documentation :sparkles:

> Note: This implementation does not include adding words to dictionary from context menu.

Please see our discussion in https://github.com/desktop/desktop/pull/10398. While being able to toggle spellchecking is on important piece of us being able to ship spellchecking in Desktop this we also need to get a handle on how to surface suggestions to the user, how to let them ignore common false positives etc. If that's something you'd like to work on we'd be happy to take a look at it!

Hovemeyer5 committed:
Add Spell Check Context Menu Options

Hovemeyer5 commented:
Added Spell Check Context Menu Options:
1. Can now also enable commit spell check from summary and description context menu.
![image](https://user-images.githubusercontent.com/7199430/99157807-5fdbee00-269a-11eb-9564-9e07539fbc34.png)

2. Can disable commit spell check from summary and description context menu.
![image](https://user-images.githubusercontent.com/7199430/99157833-8dc13280-269a-11eb-81ae-7d566a7dd9e0.png)

3. Receive suggestions
![image](https://user-images.githubusercontent.com/7199430/99157845-af221e80-269a-11eb-9f8d-25b11ab449c3.png)

4. Add to Dictionary
![image](https://user-images.githubusercontent.com/7199430/99157854-bcd7a400-269a-11eb-93ce-745d941cb0b2.png)

5. Also does not impact other custom context menus.
![image](https://user-images.githubusercontent.com/7199430/99157876-e395da80-269a-11eb-843b-05c987df6254.png)
![image](https://user-images.githubusercontent.com/7199430/99157871-d8db4580-269a-11eb-9d37-267fec7fd92d.png)
![image](https://user-images.githubusercontent.com/7199430/99157882-f0b2c980-269a-11eb-9f13-57bbf0d2448e.png)

Hovemeyer5 committed:
Merge pull request #1 from desktop/development

Update Develop

Hovemeyer5 committed:
Merge branch 'development' into optional-commit-spell-checking

Hovemeyer5 committed:
Commit Spell Check Enable Key

Hovemeyer5 committed:
Use Commit Spell Check Default

Hovemeyer5 committed:
prettier on dispatch after merge

Hovemeyer5 committed:
Use getCurrentWebContents?

Hovemeyer5 committed:
Add Mock for getCurrentWebContents

sarim commented:
After clicking add to dictionary can the user see the list of custom words added to dictionary somewhere? Maybe i mistakenly added a wrong word to dictionary and want to revert it.
Also It'd be nice to have a feature to add regex patterns to ignore spell check.  It'd help in commit messages, let me quote a comment from parent issue

> For example, if I type “Refactor FooGlobber”, I don’t want “FooGlobber” to be marked as misspelled. Nor “refactor”, which is not in all dictionaries. The same with file paths – if I write “changed src/core/index.js”, I don’t want “src” to be marked as misspelled.

With regex it'd be very easy to capture `src/core/index.js` part and ignore it.

niik committed:
Merge branch 'development' into pr/10996

niik committed:
Support deferred context menus with spelling suggestions

 showContextualMenu can now defer actually showing the context menu until receiving word from Electron that  spelling suggestions have been considered for the context menu event.

Co-Authored-By: tidy-dev <75402236+tidy-dev@users.noreply.github.com>

niik committed:
Ooops forgot

Co-Authored-By: tidy-dev <75402236+tidy-dev@users.noreply.github.com>

niik committed:
Use new deferred context menu option in commit message component

Co-Authored-By: tidy-dev <75402236+tidy-dev@users.noreply.github.com>

niik committed:
Don't need this any more

Co-Authored-By: tidy-dev <75402236+tidy-dev@users.noreply.github.com>

niik committed:
Don't need currentContextMenuItems on app store any more

Co-Authored-By: tidy-dev <75402236+tidy-dev@users.noreply.github.com>

niik committed:
:art: cleanup

tidy-dev committed:
Add platform wording and clean up to function

tidy-dev committed:
Absolute Check

tidy-dev committed:
Add documentation and more cleanup

niik commented:
@sergiou87 No rush but it'd be great if you could review this as both me and @Hovemeyer5 / @tidy-dev have touched it 🙇

sergiou87 reviewed (dismissed):
It looks great to me, great job!!! 😍

tidy-dev committed:
Remove Spellcheck Option from Preferences

sergiou87 commented on app/src/ui/lib/text-box.tsx:
```diff
@@ -288,6 +291,7 @@ export class TextBox extends React.Component<ITextBoxProps, ITextBoxState> {
           onKeyDown={this.onKeyDown}
           tabIndex={this.props.tabIndex}
           onContextMenu={this.onContextMenu}
+          spellCheck={this.props.spellcheck === true}
```
~~As discussed in Slack, seems like the changes in this file shouldn't be needed (since `TextBox` is not used in the commit message box).~~

**UPDATE:** Nope, this is needed to prevent spellchecking in other places in the app.

sergiou87 reviewed (changes requested):
~~Aside from removing some changes we discussed in Slack~~, I wasn't able to make this work (basically I don't see the red line under misspelled words, and I don't get suggestions either).

I'll take a look tomorrow and see if there is anything wrong in my machine :thinking:

**UPDATE:** found what was wrong… I needed to rebuild the app to make it work!

sergiou87 reviewed (approved):
After some testing, I found out the changes in `TextBox` are actually necessary, otherwise you get spellchecking everywhere in the app 😂 

Sorry about that!! I think this can be merged already 😌  Great job!!! 😍

niik commented:
![yesyesyes](https://user-images.githubusercontent.com/634063/105163190-e4704000-5b13-11eb-9b07-fec1f448d6a4.gif)
